### ISSUE TITLE
Parametrized module and rexport cause internal error

### ISSUE THREAD

**Issue by marco-vassena** [agda/agda#1679]
The combination of parametrized modules and the re-export crashes agda.

A first module M1 is parametrized by (A : Set) and re-exports some common modules
from the standard library.

module M1 (A : Set) where

open import Data.List public -- One public import is ok
open import Data.Unit public -- But the second causes an internal error

The first module type-checks correctly.

A second module M2 is parametrized by (B : Set) and imports  M1 instantiated with B.

module M2 (B : Set) where

open import M1 B

Type-checking M2 fails with an internal error:
An internal error has occurred. Please report this as a bug.
Location of the error: src/full/Agda/TypeChecking/Substitute.hs:343

Even more strangely the failure does not occur with only a single re-export, but only when
at least two modules are re-exported.

 Agda version 2.4.3

**Comment by marco-vassena** [comment]
It actually does not matter that the second module is parametrized.

Same behaviour for

module M2 where

postulate B : Set

open import M1 B

**Comment by andreasabel** [comment]
A shorter test case:

``` agda
postulate B : Set

module M1 (A : Set) where
  open import Data.Unit public

open M1 B

-- An internal error has occurred. Please report this as a bug.
-- Location of the error: src/full/Agda/TypeChecking/Substitute.hs:343
```

**Comment by andreasabel** [comment]
Here is a test case without `import`:

``` agda
postulate B : Set

module M where
  record ⊤ : Set where

module P (A : Set) where
  open M public

open P B
```

**Comment by andreasabel** [comment]
Agda crashes when pi-applying `M.T` to `B`.  This is an instance of #892.

**Comment by andreasabel** [comment]
@vlopezj : For the collection of module system bugs.

**Comment by andreasabel** [comment]
`git bisect` claims this bug was introduced by the fix of #892. https://github.com/agda/agda/commit/7d3036bb9640720973f7acf94302babb339dd90c

```
commit 7d3036bb9640720973f7acf94302babb339dd90c
Author: Ulf Norell <ulf.norell@gmail.com>
Date:   Thu Aug 28 13:34:16 2014 +0200

    [ issue892 ] fixed

    Required a slight change in the semantics of module application.
    Previously re-exported functions were not redefined in the new module,
    but simply aliased by the scope checker. This is wrong as issue 892
    shows, so now all exported functions are redefined, which may lead to
    ambiguous identifier errors in code that previously checked.
```

**Comment by andreasabel** [comment]
See also #1999.
